File: AngularJS/src/org/angular2/codeInsight/template/Angular2TemplateScope.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.codeInsight.template;

import com.intellij.psi.ResolveResult;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;

public abstract class Angular2TemplateScope {

  @Nullable
  private final Angular2TemplateScope myParent;
  private final List<Angular2TemplateScope> children = new ArrayList<>();

  /**
   * A scope can be created with parent scope, which contents will be included in the resolution.
   * See {@link Angular2TemplateScope#resolveAllScopesInHierarchy}
   */
  protected Angular2TemplateScope(@Nullable Angular2TemplateScope parent) {
    myParent = parent;
    if (parent != null) {
      parent.add(this);
    }
  }

  @Nullable
  public final Angular2TemplateScope getParent() {
    return myParent;
  }

  @NotNull
  public final List<Angular2TemplateScope> getChildren() {
    return Collections.unmodifiableList(children);
  }

  private void add(Angular2TemplateScope scope) {
    this.children.add(scope);
  }

  /**
   * This method is called on every provided scope and allows for providing resolve results from enclosing scopes.
   */
  public final void resolveAllScopesInHierarchy(@NotNull Consumer<? super ResolveResult> consumer) {
    Angular2TemplateScope scope = this;
    while (scope != null) {
      scope.resolve(consumer);
      scope = scope.getParent();
    }
  }

  public abstract void resolve(@NotNull Consumer<? super ResolveResult> consumer);
}

File: AngularJS/src/org/angular2/codeInsight/template/Angular2TemplateScope.kt
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.codeInsight.template

import com.intellij.psi.ResolveResult
import com.intellij.util.Processor
import java.util.*
import java.util.function.Consumer

abstract class Angular2TemplateScope
/**
 * A scope can be created with parent scope, which contents will be included in the resolution.
 * See [Angular2TemplateScope.resolveAllScopesInHierarchy]
 */
protected constructor(val parent: Angular2TemplateScope?) {
  private val children = ArrayList<Angular2TemplateScope>()

  init {
    @Suppress("LeakingThis")
    parent?.add(this)
  }

  fun getChildren(): List<Angular2TemplateScope> {
    return Collections.unmodifiableList(children)
  }

  private fun add(scope: Angular2TemplateScope) {
    this.children.add(scope)
  }

  /**
   * This method is called on every provided scope and allows for providing resolve results from enclosing scopes.
   */
  fun resolveAllScopesInHierarchy(processor: Processor<in ResolveResult>): Boolean {
    var scope: Angular2TemplateScope? = this
    var found = false
    val consumer = { resolveResult: ResolveResult ->
      if (!processor.process(resolveResult)) {
        found = true
      }
    }
    while (scope != null && !found) {
      scope.resolve(consumer)
      scope = scope.parent
    }
    return found
  }

  abstract fun resolve(consumer: Consumer<in ResolveResult>)
}
